From 5c45f3b01f6cd44831627bd808bfbf4e84b128fa Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Fri, 2 May 2014 14:36:53 -0400
Subject: [misc] mei: nfc: fix memory leak in error path

Message-id: <1399041470-21330-21-git-send-email-prarit@redhat.com>
Patchwork-id: 79308
O-Subject: [RHEL7.1 PATCH BZ 1088953 20/77] mei: nfc: fix memory leak in error path
Bugzilla: 1088953
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1088953

commit 4bff7208f332b2b1d7cf1338e50527441283a198
Author: Tomas Winkler <tomas.winkler@intel.com>
Date:   Mon Oct 21 22:05:38 2013 +0300

    mei: nfc: fix memory leak in error path

    The flow may reach the err label without freeing cl and cl_info

    cl and cl_info weren't assigned to ndev->cl and cl_info
    so they weren't freed in mei_nfc_free called on error path

    Cc: <stable@vger.kernel.org> # 3.10+
    Cc: Samuel Ortiz <sameo@linux.intel.com>
    Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/misc/mei/nfc.c b/drivers/misc/mei/nfc.c
index d0c6907..994ca4a 100644
--- a/drivers/misc/mei/nfc.c
+++ b/drivers/misc/mei/nfc.c
@@ -485,8 +485,11 @@ int mei_nfc_host_init(struct mei_device *dev)
  if (ndev->cl_info)
   return 0;
 
- cl_info = mei_cl_allocate(dev);
- cl = mei_cl_allocate(dev);
+ ndev->cl_info = mei_cl_allocate(dev);
+ ndev->cl = mei_cl_allocate(dev);
+
+ cl = ndev->cl;
+ cl_info = ndev->cl_info;
 
  if (!cl || !cl_info) {
   ret = -ENOMEM;
@@ -527,10 +530,9 @@ int mei_nfc_host_init(struct mei_device *dev)
 
  cl->device_uuid = mei_nfc_guid;
 
+
  list_add_tail(&cl->device_link, &dev->device_list);
 
- ndev->cl_info = cl_info;
- ndev->cl = cl;
  ndev->req_id = 1;
 
  INIT_WORK(&ndev->init_work, mei_nfc_init);
-- 
1.7.1